File: benchmarks/backend_crypto/reference/swift/BenchRuntime.swift
import Foundation

struct BenchMetrics {
    let warmupIterations: Int
    let measureIterations: Int
    let samples: Int
    let p50Ms: Double
    let p90Ms: Double
    let p99Ms: Double
    let meanMs: Double
    let stddevMs: Double
    let throughputOpsS: Double
    let clock: String
    let platform: String
    let perfCapability: String
    let perfCounters: String
    let perfReason: String
}

private func hostPlatform() -> String {
    #if os(macOS)
    #if arch(arm64)
    return "macos-arm64"
    #elseif arch(x86_64)
    return "macos-x64"
    #else
    return "macos-unknown"
    #endif
    #elseif os(Linux)
    #if arch(arm64)
    return "linux-arm64"
    #elseif arch(x86_64)
    return "linux-x64"
    #else
    return "linux-unknown"
    #endif
    #elseif os(Windows)
    return "windows-x64"
    #else
    return "unknown-unknown"
    #endif
}

private func percentile(_ samplesMs: [Double], _ p: Double) -> Double {
    if samplesMs.isEmpty {
        return 0.0
    }
    let idx = Int((p / 100.0) * Double(samplesMs.count - 1))
    return samplesMs[idx]
}

func runBench(benchOps: Int = 1, _ body: () -> Void) -> BenchMetrics {
    let warmup = 50
    let iters = 1000
    var samplesMs: [Double] = []
    samplesMs.reserveCapacity(iters)

    for _ in 0..<warmup {
        body()
    }
    for _ in 0..<iters {
        let t0 = DispatchTime.now().uptimeNanoseconds
        body()
        let t1 = DispatchTime.now().uptimeNanoseconds
        samplesMs.append(Double(t1 - t0) / 1_000_000.0)
    }

    samplesMs.sort()
    let sumMs = samplesMs.reduce(0.0, +)
    let meanMs = samplesMs.isEmpty ? 0.0 : (sumMs / Double(samplesMs.count))
    let variance: Double
    if samplesMs.count <= 1 {
        variance = 0.0
    } else {
        variance = samplesMs.reduce(0.0) { partial, sample in
            let delta = sample - meanMs
            return partial + delta * delta
        } / Double(samplesMs.count - 1)
    }
    let totalS = sumMs / 1000.0
    let throughput = totalS > 0.0 ? (Double(iters * benchOps) / totalS) : 0.0

    return BenchMetrics(
        warmupIterations: warmup,
        measureIterations: iters,
        samples: samplesMs.count,
        p50Ms: percentile(samplesMs, 50.0),
        p90Ms: percentile(samplesMs, 90.0),
        p99Ms: percentile(samplesMs, 99.0),
        meanMs: meanMs,
        stddevMs: variance.squareRoot(),
        throughputOpsS: throughput,
        clock: "steady_clock",
        platform: hostPlatform(),
        perfCapability: "unsupported",
        perfCounters: "unsupported",
        perfReason: "not_implemented"
    )
}

func emitMetrics(_ metrics: BenchMetrics) {
    print("{\"warmup_iterations\":\(metrics.warmupIterations),\"measure_iterations\":\(metrics.measureIterations),\"samples\":\(metrics.samples),\"p50_ms\":\(metrics.p50Ms),\"p90_ms\":\(metrics.p90Ms),\"p99_ms\":\(metrics.p99Ms),\"mean_ms\":\(metrics.meanMs),\"stddev_ms\":\(metrics.stddevMs),\"throughput_ops_s\":\(metrics.throughputOpsS),\"clock\":\"\(metrics.clock)\",\"platform\":\"\(metrics.platform)\",\"perf_capability\":\"\(metrics.perfCapability)\",\"perf_counters\":\"\(metrics.perfCounters)\",\"perf_reason\":\"\(metrics.perfReason)\"}")
}

File: benchmarks/backend_crypto/reference/rust/bench_runtime.rs
use std::time::Instant;

pub struct BenchMetrics {
    pub warmup_iterations: usize,
    pub measure_iterations: usize,
    pub samples: usize,
    pub p50_ms: f64,
    pub p90_ms: f64,
    pub p99_ms: f64,
    pub mean_ms: f64,
    pub stddev_ms: f64,
    pub throughput_ops_s: f64,
    pub clock: &'static str,
    pub platform: &'static str,
    pub perf_capability: &'static str,
    pub perf_counters: &'static str,
    pub perf_reason: &'static str,
}

fn host_platform() -> &'static str {
    if cfg!(target_os = "macos") && cfg!(target_arch = "aarch64") {
        "macos-arm64"
    } else if cfg!(target_os = "macos") && cfg!(target_arch = "x86_64") {
        "macos-x64"
    } else if cfg!(target_os = "linux") && cfg!(target_arch = "aarch64") {
        "linux-arm64"
    } else if cfg!(target_os = "linux") && cfg!(target_arch = "x86_64") {
        "linux-x64"
    } else if cfg!(target_os = "windows") && cfg!(target_arch = "x86_64") {
        "windows-x64"
    } else {
        "unknown-unknown"
    }
}

fn percentile(samples_ms: &[f64], p: f64) -> f64 {
    if samples_ms.is_empty() {
        return 0.0;
    }
    let idx = ((p / 100.0) * ((samples_ms.len() - 1) as f64)) as usize;
    samples_ms[idx]
}

pub fn run_bench<F>(mut f: F, bench_ops: usize) -> BenchMetrics
where
    F: FnMut(),
{
    let warmup = 50usize;
    let iters = 1000usize;
    let mut samples_ms = Vec::with_capacity(iters);

    for _ in 0..warmup {
        f();
    }
    for _ in 0..iters {
        let started = Instant::now();
        f();
        let elapsed_ms = started.elapsed().as_secs_f64() * 1000.0;
        samples_ms.push(elapsed_ms);
    }
    samples_ms.sort_by(|lhs, rhs| lhs.partial_cmp(rhs).unwrap());

    let sum_ms: f64 = samples_ms.iter().sum();
    let mean_ms = if samples_ms.is_empty() {
        0.0
    } else {
        sum_ms / (samples_ms.len() as f64)
    };
    let variance = if samples_ms.len() <= 1 {
        0.0
    } else {
        samples_ms
            .iter()
            .map(|sample| {
                let delta = *sample - mean_ms;
                delta * delta
            })
            .sum::<f64>()
            / ((samples_ms.len() - 1) as f64)
    };
    let total_s = sum_ms / 1000.0;
    let throughput = if total_s > 0.0 {
        ((iters * bench_ops) as f64) / total_s
    } else {
        0.0
    };

    BenchMetrics {
        warmup_iterations: warmup,
        measure_iterations: iters,
        samples: samples_ms.len(),
        p50_ms: percentile(&samples_ms, 50.0),
        p90_ms: percentile(&samples_ms, 90.0),
        p99_ms: percentile(&samples_ms, 99.0),
        mean_ms,
        stddev_ms: variance.sqrt(),
        throughput_ops_s: throughput,
        clock: "steady_clock",
        platform: host_platform(),
        perf_capability: "unsupported",
        perf_counters: "unsupported",
        perf_reason: "not_implemented",
    }
}

pub fn emit_metrics_json(metrics: &BenchMetrics) {
    println!(
        "{{\"warmup_iterations\":{},\"measure_iterations\":{},\"samples\":{},\"p50_ms\":{},\"p90_ms\":{},\"p99_ms\":{},\"mean_ms\":{},\"stddev_ms\":{},\"throughput_ops_s\":{},\"clock\":\"{}\",\"platform\":\"{}\",\"perf_capability\":\"{}\",\"perf_counters\":\"{}\",\"perf_reason\":\"{}\"}}",
        metrics.warmup_iterations,
        metrics.measure_iterations,
        metrics.samples,
        metrics.p50_ms,
        metrics.p90_ms,
        metrics.p99_ms,
        metrics.mean_ms,
        metrics.stddev_ms,
        metrics.throughput_ops_s,
        metrics.clock,
        metrics.platform,
        metrics.perf_capability,
        metrics.perf_counters,
        metrics.perf_reason
    );
}
